src/org/sosy_lab/java_smt/solvers/cvc5/CVC5InterpolatingProver.java

```diff
@@ -0,0 +1,334 @@
+// This file is part of JavaSMT,
+// an API wrapper for a collection of SMT solvers:
+// https://github.com/sosy-lab/java-smt
+//
+// SPDX-FileCopyrightText: 2023 Dirk Beyer <https://www.sosy-lab.org>
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package org.sosy_lab.java_smt.solvers.cvc5;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import io.github.cvc5.Kind;
+import io.github.cvc5.Term;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import org.sosy_lab.common.ShutdownNotifier;
+import org.sosy_lab.java_smt.api.BooleanFormula;
+import org.sosy_lab.java_smt.api.FormulaManager;
+import org.sosy_lab.java_smt.api.InterpolatingProverEnvironment;
+import org.sosy_lab.java_smt.api.SolverContext.ProverOptions;
+import org.sosy_lab.java_smt.api.SolverException;
+
+public class CVC5InterpolatingProver extends CVC5AbstractProver<Term>
+    implements InterpolatingProverEnvironment<Term> {
+
+  private final FormulaManager mgr;
+  // private final CVC5FormulaCreator creator;
+  private final Set<Term> assertedFormulaHash = new HashSet<>(); // comparable to SMTInterpols
```

Comment:
Is a plain Set sufficient?

Counterexample:
What happens on the following process: "push(A), push(B), push(A), pop()"? Then "A" is no longer in the set.

Reply:
I swapped it with a `Deque` in the newest commit. It should solve it, right?

Reply:
No. A one-dimensional data-structure will not help here.
If you really need to keep track of the existing levels, then you might need use something like a Deque<Collection<Term>>, as it is also done in some other ProverEnvironments.